### ISSUE TITLE
Cleanup the deduplication in materialized views.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#60008]
Remove these settings:

```
deduplicate_blocks_in_dependent_materialized_views
update_insert_deduplication_token_in_dependent_materialized_views
```

They will become no-ops.

Instead, the deduplication should work in a similar way to when an explicit `insert_deduplication_token` is specified,
but using the hash of the inserted blocks of data as a token.

**How an INSERT query works:**

1. Each INSERT from a user is represented as a stream of data. This stream of data is represented by a sequence of one or more blocks (`max_insert_block_size`). These blocks can be squashed together to form larger blocks (`min_insert_block_size_rows`, `min_insert_block_size_bytes`), which also represent a sequence of one or more blocks. The blocks can also be formed by the data sent from different concurrent clients, which is the case of `async_insert`.

2. Each block from a sequence is processed by the table (`IStorage::write` method), either sequentially or by multiple blocks in parallel (`max_insert_threads`), and by materialized views attached to this table, and all materialized views in the chain. Each materialized view does some processing on each block and writes it to its underlying table using the (`IStorage::write` method).

3. The table's write method splits each block by partition key, forming smaller blocks. These blocks are then committed (one by one) atomically to Keeper, and to the filesystem, and to the state of the table in memory. During the atomic commit of each resulting block, it is checked for duplicates, using a set of hashes stored in Keeper.

**How a deduplication token works:**

A user can provide a deduplication token along with an INSERT query using the `insert_deduplication_token` setting.

1. This token is combined with the deterministic sequential number of a block received by INSERT, possibly after squashing (as the squashing process is also deterministic), but before combining the blocks from different clients in the case of `async_insert` (in this case, the tokens will correspond to the ranges in the combined block). 

2. Then it is combined with the table identifier for each participating table in the chain of materialized views (the table identifier is immutable and persistent UUID for `Atomic` and `Replicated` databases; or represented by the database and table name in the obsolete `Ordinary` databases). Then it is combined with the partition identifier of every table's partition. Finally, this combination is hashed to make a hash, and this hash (or multiple hashes in the case of `async_insert`) is matched to the set of hashes in Keeper.

**How the deduplication should work without a deduplication token:**

Calculate a hash of every block (range of block in the case of `async_insert`) received by INSERT after step 1, and use it in the same way as an explicit deduplication token in the later steps.

**Benefits**

1. Insertion into a table with materialized views becomes idempotent (it is safe to repeat the insertion in the case of partial errors or any other case).
2. We take out the calculation of block hash from the MergeTree engine, and it becomes unified with the way how we do it for `async_insert`.
3. The hash will be calculated for inserts into all table engines and can be reused by other table engines, such as `Memory`, or even `Distributed`. However, we can omit this calculation if there are no tables supporting deduplication in the chain.
4. We will remove obscure settings `deduplicate_blocks_in_dependent_materialized_views` and `update_insert_deduplication_token_in_dependent_materialized_views`, that cannot be used by anyone except their authors.

**Comment by UnamedRus** [comment]
> Then it is combined with the table identifier for each participating table in the chain of materialized views (the table identifier is immutable and persistent UUID for Atomic and Replicated databases; or represented by the database and table name in the obsolete Ordinary databases). 


People can have tables with the different database/table, but they still share zookeeper path. (circular replication for ex, it's not recommended but still)
Same goes about UUID, people could create table without ON CLUSTER statement, but they still point to the same (manually defined) zookeeper path.

So, it's better to use zookeeper_path instead.

**Comment by CheSema** [comment]
That comment is just for me as a note:

- MT and RMT have different way to store deduplication log. 
  - MT does it with local disk.
  - RMT doest with keeper

- I suspect that `finishDelayedChunk()` is called 2 times more than needed. see `max_insert_delayed_streams_for_parallel_write`.

**Comment by tavplubix** [comment]
> a lot of part's actions do deduplication_log->dropPart(part->info);. After that an duplicate could be inserted.

This is expected, it's the same in RMT, see ``clearBlocksInPartition

**Comment by CheSema** [comment]
I have noticed that `deduplicate_blocks_in_dependent_materialized_views` works only for `ReplicatedMergeTree` engine. For non replicated `MergeTree` deduplication are defined only by merge tree tables settings `non_replicated_deduplication_window`. 
For `ReplicatedMergeTree` the deduplication could be on/off by session/query level setting `deduplicate_blocks_in_dependent_materialized_views`.

Firstly I have thought that is a bug and I should fix that. But I have change my mind.
I want to deprecate `deduplicate_blocks_in_dependent_materialized_views` at the following PR. Deduplication is on a property on session/query level. It should be on table settings level. You enable `deduplication_log` and set `*_deduplication_window` for the table. That should work for materialized views as well.

**Comment by den-crane** [comment]
git or blockchain.

*    dedup_checksum = parent_checksum + checksum(of a current block)
*    the first parent_checksum is checksum of the inserted block or token
*    can be several parents if MatView squashes a block from several source blocks.

**Comment by CheSema** [comment]
Please check this https://github.com/ClickHouse/ClickHouse/pull/61601
I have totaly reworked how it works. I will soon add description what exactly I did.

> Also maybe it has sense to include the name of MV into a checksum, for complicated (branched) scenarios demonstrated by jrdi.

Yes, it would work like that.

> deduplicate_blocks_in_dependent_materialized_views is needed now because narrow MV can produce exactly the same block for different source inserts, thus Clickhouse will skip (deduplicate) valid inserts into MV table.

`block_id` is constructed now in a way that it depends from source and from MV it self. Different inserts are never deduplicated in MV even if data is the same binary.

**Comment by CheSema** [comment]
> can be several parents if MatView squashes a block from several source blocks.

I do not see such possibility. Inner query MatView always works with one block. That block could be i'th in a series of blocks from source. A result of Inner query execution could be a series of blocks as well.
But never MatView get input from different sources.

**Comment by den-crane** [comment]
@CheSema Maybe I said it incorrectly.

Here is an example:

```
create table A (A Int64 ) Engine=MergeTree order by A;
create table B (C Int64) Engine=MergeTree order by tuple();
create materialized view B_mv to B as select count() C from A;
insert into A select * from numbers(2e6);

select distinct _part from A;
   ┌─_part─────┐
1. │ all_2_2_0 │
   └───────────┘
   ┌─_part─────┐
2. │ all_1_1_0 │
   └───────────┘

select distinct _part from B;
   ┌─_part─────┐
1. │ all_1_1_0 │
   └───────────┘
```

A table formed 2 parts / 2 checksums
B table formed 1 part / 1 checksum

**Comment by CheSema** [comment]
Oh, actually you pointed to a bug in my PR. Thanks. Will handle it.

**Comment by CheSema** [comment]
Here is an interesting case.

https://github.com/ClickHouse/ClickHouse/pull/61601/files#r1626339684

I want to discuss that case.

schema
```
CREATE TABLE partitioned_table
    (key Int64, value String)
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{database}/03008_deduplication_insert_into_partitioned_table', '{replica}')
    partition by key % 10
    order by tuple();

CREATE MATERIALIZED VIEW mv_table (key Int64, value String)
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{database}/03008_deduplication_insert_into_partitioned_table_mv', '{replica}')
    ORDER BY tuple()
    AS SELECT key, value FROM partitioned_table;
```

inserts
```
INSERT INTO partitioned_table VALUES (1, 'A'), (2, 'B');
INSERT INTO partitioned_table VALUES (1, 'A'), (2, 'C');
INSERT INTO partitioned_table VALUES (1, 'D'), (2, 'B');
```

results
```

SELECT 'partitioned_table is deduplicated bacause deduplication works in scope of one partiotion:';
SELECT * FROM partitioned_table ORDER BY ALL;
SELECT 'mv_table is not deduplicated because the inserted blocks was different:';
SELECT * FROM mv_table ORDER BY ALL;
```
are 
```
+partitioned_table:
+1	A
+1	D
+2	B
+2	C
+mv_table:
+1	A
+1	A
+1	D
+2	B
+2	B
+2	C
```

**Comment by alexey-milovidov** [comment]
Good example, I understand it!
Yes, it is explainable. Some alternatives are possible:

1. Push to materialized views after splitting to partitions (but it will complicate the logic; it will require the INSERT query pipeline to look into the table).
2. Perform deduplication before splitting into partitions (but this changes the way how we work with metadata in ZooKeeper, and also makes the logic of the number of hashes to keep different).

I don't like any of these alternatives. Let's keep it as it is.

**Comment by alexey-milovidov** [comment]
This has been implemented.
The second part for deduplication of asynchronous inserts in materialized views is out of scope of this task.
